easybuild/tools/modules.py

```diff
@@ -742,14 +742,44 @@ def modpath_extensions_for(self, mod_names):
             r'^\s*module\s+use\s+"?([^"\s]+)"?',  # 'module use' in Tcl module files
             r'^\s*prepend-path\s+MODULEPATH\s+"?([^"\s]+)"?',  # prepend to $MODULEPATH in Tcl modules
             r'^\s*prepend_path\(\"MODULEPATH\",\s*\"(\S+)\"',  # prepend to $MODULEPATH in Lua modules
+            r'^\s*prepend_path\(\"MODULEPATH\",\s*pathJoin\((.+)\)\)',  # prepend to $MODULEPATH in Lua modules using pathJoin
         ])
         modpath_ext_regex = re.compile(modpath_ext_regex, re.M)
 
         modpath_exts = {}
         for mod_name in mod_names:
             modtxt = self.read_module_file(mod_name)
-            exts = [ext for tup in modpath_ext_regex.findall(modtxt) for ext in tup if ext]
+            matches = modpath_ext_regex.findall(modtxt)
+            exts = [ext for tup in matches for ext in tup if ext]
             self.log.debug("Found $MODULEPATH extensions for %s: %s", mod_name, exts)
+
+            # Need to expand environment variables and join paths, e.g. when --subdir-user-modules is used
+            for i, ext in enumerate(exts):
+
+                if not matches[i][2]: # can skip plain literal Lua match
+
+                    # expand environment variables
+                    if matches[i][3]: # pathJoin match (Lua)
+                        envstr = "os.getenv("
+                    else: # TCL match
+                        envstr = "$env("
+                    envlocstart = ext.find(envstr)
+                    while envlocstart != -1:
+                        envlocend = ext.find(")", envlocstart+len(envstr))
+                        if envlocend == -1:
+                            break
+                        envvar = ext[envlocstart+len(envstr):envlocend].strip('"')
+                        envval = os.environ.get(envvar, "")
+                        ext = ext.replace(ext[envlocstart:envlocend+1], envval)
+                        envlocstart = ext.find(envstr,envlocstart+len(envval))
+
+                    # join paths (Lua)
+                    if matches[i][3]:
+                        ext = os.path.join(*ext.replace(",","").replace('"','').split())
```

Comment:
@bartoldeman There's quite a lot going on here, and it looks quite tedious to me as well.

If we go through with this, we should flesh this out in a separate function, so we can test in isolation...

```python
def interpret_lua_statement(txt):
    """Interpret raw Lua statement: resolve environment variables, join paths wher `pathJoin` is specified"""
    return ...

...

exts = []
for raw_ext in raw_exts:
    if lua:
        ext = process_raw_lua(raw_ext)
    else:
        ext = process_raw_tcl(raw_ext)
    exts.append(ext)
```

That being said, this looks... adventurous.